README.md
# Pomodoro

This was fun. My first VSCode extension and Git repository. This is my version of [Pomodoro Technique](https://en.wikipedia.org/wiki/Pomodoro_Technique) as extension for VSCode. Hope you like it.

![Pomodoro](images/pomodoro.gif)

## Features

Breaks down work into intervals (pomodoros) of 25 minutes separated by short breaks of 5 minutes, 4 pomodoros (sets) are separated by long breaks of 15 minutes. 

After you start Pomodoro (Command Palette: Pomodoro Start/Pause) you will have in statusbar `🍅 Work 1/1 25:00`:
* `🍅` : the unicode character for tomato
* `Work` : you are in the middle of a pomodoro
* `1/1` : pomodoro 1 in set 1
* `25:00` : the timer, after a second will become 24:59 and so on...

Above are the defaults that can be changed from settings.

The extension creates items in:
* Command Palette: Pomodoro Start/Pause and Pomodoro Reset
* Statusbar (and its tooltip): with custom texts from settings
* Notifications: with custom texts from settings
* Settings: Available settings bellow


## Extension Settings

This extension contributes the following settings:

* `pomodoro.work` : time of a pomodoro
* `pomodoro.break` : time of break between pomodoros
* `pomodoro.repeat` : how many pomodoros in a set
* `pomodoro.longBreak` : time of break between sets
* `pomodoro.autoStart` : auto start on VSCode startup
* `pomodoro.filePath` : send changed state `{ repeat: number, task: number, set: number, is: string, active: string }` to your file `module.exports = state => console.log(state)`
* `pomodoro.make.notification` : display notifications
* `pomodoro.make.statusbar:{alignment;priority}` : create a statusbar item
* `pomodoro.texts:{notification;statusbar;tooltip}` : the texts displayed, available placeholders with their outputs (you can see in the default values how they work):
  * `%work{Work}` : Work
  * `%break{Break}` : Break
  * `%repeat` : 4 
  * `%longBreak{Long Break}` : Long Break
  * `%started{Started}` : Started
  * `%paused{Paused}` : Paused
  * `%reset{Reset}` : Reset
  * `%task` : 1,2,3,4,1,2...
  * `%tasksLeft` : 4,3,2,1,4,3...
  * `%set` : 1,2,3,4,5,6...
  * `%hour` : 01,00,23,22...
  * `%minute` : 01,00,59,58...
  * `%second` : 01,00,59,58...


## Release Notes

### 1.0.0

Initial release

### 1.1.0

Remember the state when opening a folder, reloading a window, etc

-----------------------------------------------------------------------------------------------------------

[Buy me a Coke](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=JCKTBQ6RWCBBA)


**Enjoy!**


CHANGELOG.md
# Change Log
All notable changes to the "Pomodoro" extension will be documented in this file.

## [1.1.0] - 2018-08-25
### Added
Remember the state when opening a folder, reloading a window, etc


src/pomodoro.ts
'use strict'
import * as vscode from 'vscode'
export class Pomodoro {
  second = 1000 
  minOrSec = 60
  work = 25
  break = 5
  longBreak = 30
  currentTime = this.work * this.second * this.minOrSec
  repeat = 4
  tasksLeft = this.repeat
  autoStart = false
  set = 1
  task = 1
  is = 'reset'
  nextIs = 'start'
  active = ''
  currentHour = '00'
  currentMinute = '00'
  currentSecond = '00'
  texts = { statusbar: '', notification: '', tooltip: '' }
  vsCodeStatusBar: vscode.StatusBarItem | { command: string; show: () => void; text: string; tooltip: string; dispose: () => void; hide: () => void} = { command: '', show: () => {}, dispose: () => {}, hide: () => {}, text: '', tooltip: '' }
  make = { statusbar: { priority: 1000, alignment: 'right' }, notification: true }
  file: any = null
  setInterval: any = null
  context: vscode.ExtensionContext
  constructor (context: vscode.ExtensionContext) {
    this.context = context
    this.stateGetConf()
    this.stateGet()
    this.setup()    
    if (this.autoStart && this.is === 'reset') this.startPause()
    else if (this.is === 'started') this.tick()
  }
  startPause(): void {
    if (this.nextIs === 'pause') {
      clearInterval(this.setInterval)
      this.setInterval = null
      this.is = 'paused'
      this.nextIs = 'resume'
      this.textNotification()
      this.textStatusBar()
      this.textTooltip()
      this.sendToFile()
      this.stateUpdate()
    } else if (this.nextIs === 'resume') {
      this.is = 'started'
      this.nextIs = 'pause'
      this.textNotification()
      this.textStatusBar()
      this.textTooltip()
      this.sendToFile()
      this.tick()
    } else {
      this.is = 'started'
      this.nextIs = 'pause'
      this.active = 'work'
      this.HMS()
      this.textNotification()
      this.textStatusBar()
      this.textTooltip()
      this.sendToFile()
      this.tick()
    }
  }
  reset(): void {
    clearInterval(this.setInterval)
    this.setInterval = null
    this.set = 1
    this.task = 1
    this.tasksLeft = this.repeat
    this.is = 'reset'
    this.nextIs = 'start'
    this.active = ''
    this.currentTime = this.work * this.second * this.minOrSec
    this.HMS()
    this.textNotification()
    this.textStatusBar()
    this.textTooltip()
    this.vsCodeStatusBar.show()
    this.sendToFile()
    this.stateUpdate()
  }
  setup(): void {
    const config = vscode.workspace.getConfiguration('pomodoro')
    this.file = config.filePath && this.preRequire(config.filePath, () => vscode.window.showWarningMessage('Pomodoro error. Check the file path in `pomodoro.filePath` setting.')) ? require(config.filePath) : null
    let updateTime = false
    const newMinOrSec = config.interval === 'seconds' ? 1 : 60
    if (newMinOrSec !== this.minOrSec) {
      updateTime = true
      this.minOrSec = newMinOrSec
    }
    if (this.work !== config.work) {
      updateTime = true
      this.work = config.work
    }
    if (this.break !== config.break) {
      updateTime = true
      this.break = config.break
    }
    if (this.longBreak !== config.longBreak) {
      updateTime = true
      this.longBreak = config.longBreak
    }
    let updateStateConf = false
    if (updateTime) {
      this.currentTime = this.work * this.second * this.minOrSec
      updateStateConf = true
    }
    if (this.repeat !== config.repeat) {
      this.repeat = this.tasksLeft = config.repeat
      this.task = 1
      updateStateConf = true
    }
    if (updateStateConf) this.stateUpdateConf()
    this.texts = config.texts
    this.make = config.make
    if (this.make.statusbar) {
      if (this.vsCodeStatusBar.command) {
        this.vsCodeStatusBar.hide()
        this.vsCodeStatusBar.dispose()
      }
      this.vsCodeStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment[this.make.statusbar.alignment === 'left' ? 'Left' : 'Right'], this.make.statusbar.priority)
      this.vsCodeStatusBar.command = 'extension.pomodoroStartPause'
      this.HMS()
      this.textStatusBar()
      this.textTooltip()
      this.vsCodeStatusBar.show()
    }
    this.autoStart = config.autoStart
  }
  tick(): void {
    this.setInterval = setInterval(() => {
      this.currentTime -= this.second
      this.HMS()
      if (this.currentTime <= 0) {
        if (this.active === 'longBreak') {
          this.active = 'work'
          this.set++
          this.task = 1
          this.tasksLeft = this.repeat
          this.currentTime = this.work * this.second * this.minOrSec
        } else if (this.active === 'break') {
          this.active = 'work'
          this.task++
          this.tasksLeft--
          this.currentTime = this.work * this.second * this.minOrSec
        } else {
          if (this.task < this.repeat) {
            this.active = 'break'
            this.currentTime = this.break * this.second * this.minOrSec
          } else {
            this.active = 'longBreak'
            this.currentTime = this.longBreak * this.second * this.minOrSec
          }
        }
        this.textNotification()
        this.sendToFile()
      }
      this.textStatusBar()
      this.stateUpdate()
    }, this.second)
  }
  textStatusBar(): void {
    if (!this.make.statusbar) return
    this.vsCodeStatusBar.text = this.reg(this.texts.statusbar)
  }
  textTooltip(): void {
    if (!this.make.statusbar) return
    this.vsCodeStatusBar.tooltip = this.reg(this.texts.tooltip)
  }
  textNotification(): void {
    if (!this.make.notification) return
    vscode.window.showInformationMessage(this.reg(this.texts.notification))
  }
  reg(text: string): string {
    return text
      .replace(/%work{(.*?)}|%break{(.*?)}|%longBreak{(.*?)}/gm,
        this.active === 'work'
          ? '$1'
          : this.active === 'break'
            ? '$2'
            : this.active === 'longBreak'
              ? '$3'
              : '')
      .replace(/%started{(.*?)}|%paused{(.*?)}|%reset{(.*?)}/gm,
        this.is === 'started'
          ? '$1'
          : this.is === 'paused'
            ? '$2'           
            : this.is === 'reset'
              ? '$3'
              : '')
      .replace(/%repeat/gm, String(this.repeat))
      .replace(/%tasksLeft/gm, String(this.tasksLeft))
      .replace(/%task/gm, String(this.task))
      .replace(/%set/gm, String(this.set))
      .replace(/%hour/gm, this.currentHour)
      .replace(/%minute/gm, this.currentMinute)
      .replace(/%second/gm, this.currentSecond)
  }
  sendToFile(): void {
    if (!this.file) return
    this.file({
      repeat: this.repeat,
      task: this.task,
      set: this.set,
      is: this.is,
      active: this.active
    })
  }
  HMS(): void {
    let seconds = this.currentTime / this.second
    const hours = Math.trunc(seconds / 3600) % 24
    seconds = seconds % 3600
    const minutes = Math.trunc(seconds / 60)
    seconds = seconds % 60
    this.currentHour = ('0' + String(hours)).slice(-2)
    this.currentMinute = ('0' + String(minutes)).slice(-2)
    this.currentSecond = ('0' + String(seconds)).slice(-2)
  }
  stateUpdate(): void {
    this.context.globalState.update('pomodoroState', {
      is: this.is,
      task: this.task,
      set: this.set,
      tasksLeft: this.tasksLeft,
      nextIs: this.nextIs,
      active: this.active,
      currentTime: this.currentTime
    })
  }
  stateGet(): void {
    const state = this.context.globalState.get('pomodoroState', {
      is: this.is,
      task: this.task,
      set: this.set,
      tasksLeft: this.tasksLeft,
      nextIs: this.nextIs,
      active: this.active,
      currentTime: this.work * this.second * this.minOrSec
    })
    this.is = state.is
    this.task = state.task
    this.set = state.set
    this.tasksLeft = state.tasksLeft
    this.nextIs = state.nextIs
    this.active = state.active
    this.currentTime = state.currentTime
  }
  stateUpdateConf(): void {
    this.context.globalState.update('pomodoroConf', {
      repeat: this.repeat,
      minOrSec: this.minOrSec,
      work: this.work,
      break: this.break,
      longBreak: this.longBreak,
    })
  }
  stateGetConf(): void {
    const pomodoroConf = this.context.globalState.get('pomodoroConf', {
      repeat: this.repeat,
      minOrSec: this.minOrSec,
      work: this.work,
      break: this.break,
      longBreak: this.longBreak,
    })
    this.repeat = pomodoroConf.repeat
    this.minOrSec = pomodoroConf.minOrSec
    this.work = pomodoroConf.work
    this.break = pomodoroConf.break
    this.longBreak = pomodoroConf.longBreak
  }
  preRequire(path = '', onNotFound = (): void => {}): boolean {
    try {
      require.resolve(path)
    } catch (e) {
      onNotFound()
      return false
    }
    return true
  }
}
